pkg/apis/pipeline/v1beta1/task_conversion.go

```diff
@@ -139,3 +177,138 @@ func (ts *TaskSpec) ConvertFrom(ctx context.Context, source *v1.TaskSpec) error
 	ts.Description = source.Description
 	return nil
 }
+
+// taskDeprecation contains deprecated fields of a Task
+type taskDeprecation struct {
+	// DeprecatedSteps contains Steps of a Task that with deprecated fields defined.
+	// +listType=atomic
+	DeprecatedSteps []Step `json:"deprecatedSteps,omitempty"`
+	// DeprecatedStepTemplate contains stepTemplate of a Task that with deprecated fields defined.
+	DeprecatedStepTemplate *StepTemplate `json:"deprecatedStepTemplate,omitempty"`
+}
+
+// taskDeprecations contains deprecated fields of Tasks that belong to the same Pipeline or PipelineRun
+// the key is Task name
+type taskDeprecations map[string]taskDeprecation
+
+// serializeTaskDeprecations appends the current Task's deprecation info to annotation of the object.
+// The object could be Task, TaskRun, Pipeline or PipelineRun
+func serializeTaskDeprecations(meta *metav1.ObjectMeta, spec *TaskSpec, taskName string) error {
+	var taskDeprecation *taskDeprecation
+	if spec.HasDeprecatedFields() {
+		taskDeprecation = retrieveTaskDeprecation(spec)
+	}
+	var existingDeprecations = taskDeprecations{}
+	if str, ok := meta.Annotations[TaskDeprecationsAnnotationKey]; ok {
+		if err := json.Unmarshal([]byte(str), &existingDeprecations); err != nil {
+			return fmt.Errorf("error deserializing key %s from metadata: %w", TaskDeprecationsAnnotationKey, err)
+		}
+	}
+	if taskDeprecation != nil {
+		existingDeprecations[taskName] = *taskDeprecation
+		return version.SerializeToMetadata(meta, existingDeprecations, TaskDeprecationsAnnotationKey)
+	}
+	return nil
+}
+
+// deserializeTaskDeprecations retrieves deprecation info of the Task from object annotation.
+// The object could be Task, TaskRun, Pipeline or PipelineRun.
+func deserializeTaskDeprecations(meta *metav1.ObjectMeta, spec *TaskSpec, taskName string) error {
+	var existingDeprecations = taskDeprecations{}
+	if meta == nil || meta.Annotations == nil {
+		return nil
+	}
+	if str, ok := meta.Annotations[TaskDeprecationsAnnotationKey]; ok {
+		if err := json.Unmarshal([]byte(str), &existingDeprecations); err != nil {
+			return fmt.Errorf("error deserializing key %s from metadata: %w", TaskDeprecationsAnnotationKey, err)
```

Comment:
nit: is it possible to test the errors? like test a random string value of the TaskDeprecationsAnnotationKey?